### ISSUE TITLE
default inheritance raises error for certain orderings

### ISSUE THREAD

**Issue by asher-pembroke** [EnsembleGovServices/kamodo-core#14]
Default inheritance allows for the creation of new functions that leverage existing defaults.
However, if the new function introduces symbols for which there are no defaults, then an error is raised when forge tries to generate a signature. 

```python
try:
    k = Kamodo(f=lambda x=2: x**2)
    k['g(x,y)'] = 'y*f' # raises registration error
except SyntaxError as m:
    print(m)
```
> non-default parameter follows default parameter

**Comment by pythonguru101** [comment]
In order to leverage the default inheritance with new function in existing default you need to follow the ordering in proper manner. There should always be a non default parameter followed by a default one.
Order of defining parameter in function are :
1. positional parameter or non-default parameter i.e (a,b,c)
2. keyword parameter or default parameter i.e (a="b",r="j")
3. keyword-only parameter i.e (*args)
4. var-keyword parameter i.e (**kwargs)

In this scenario, if you try reordering the parameters the errors will be resolved just as shown below:

```python
try:
    k = Kamodo(f=lambda x=2: x**2)
    k['g(y,x)'] = 'y*f'
    print(k)
except SyntaxError as m:
    print(m)
```

**Comment by asher-pembroke** [comment]
>  if you try reordering the parameters the errors will be resolved just as shown below:

Yes, but we don't want to require the user to manually specify the order. When they write the following, kamodo should generate a signature with a correct ordering:
```python
    k = Kamodo(f=lambda x=2: x**2)
    k['g'] = 'y*f' # registers g(y, x) automatically
```
We may also want to raise a more explicit syntax error if the user tries to pick an ordering that would result in an improper function signature. Something to the effect of "Improper left-hand-side signature: default variables must follow positional variables."

**Comment by asher-pembroke** [comment]
This is what I get currently (something the end user shouldn't have to deal with)

try:
    k = Kamodo(f=lambda x=2: x**2)
    k['g'] = 'y*f' # registers g(y, x) automatically
except SyntaxError as m:
    print(m) # non-default parameter follows default parameter

**Comment by asher-pembroke** [comment]
When you make changes to kamodo core, run our test suite and make sure all tests pass. Run the tests like this:

```console
cd kamodo
pytest kamodo/test_kamodo.py kamodo/test_plotting.py  kamodo/test_utils.py
```

Most of the tests fail when I install your branch.

In addition, for this feature to work we want the following tests to pass.

```python
try:
    k = Kamodo(f=lambda x=2: x**2)
    k['g'] = 'y*f'
    
    assert k.g(3) == 3*(2**2)
    assert k.g(3,4) == 3*(4**2)
except SyntaxError as m:
    print(m) # non-default parameter follows default parameter
    raise
except AssertionError as m:
    print('assertion error') # k.g(3) != 3*(2**2)
    raise
```

However, function signature for g looks wrong:

$$f{\left(x \right)} = \lambda{\left(x \right)}$$
$$g{\left(x \right)} = y f{\left(x \right)}$$

kamodo should have detected that $g$ should be a function of y and x, with x coming last  since x is the default.

Sometimes when I try evaluatating `g`, I get an expression with symbols!

```python
k.g(3) # returns 9y
```
This seems to happen randomly, I think because `rhs_expr.free_symbols` returns a set which has an inconsistent ordering.

In kamodo, all evaluations should result in numeric answers because symbols are eliminated by substitution. The only time symbols would end up in the result would be if symbols were passed in as arguments to a function:

```python
from sympy.abc import alpha

k.f(alpha)
```
This returns the symbol `$\alpha^2$`, which is fine. (I could see some applications of this trick, and I hadn't thought of it before now. In any case, this doesn't solve our registration problem.)

Looking at your code, I can see a number of issues

```python

# This code depends on the number of arguments and only works in this special case
dict_symbol = {}
for i in range(len(rhs_expr.args[1].args)):
    dict_symbol["var{0}".format(i)] = rhs_expr.args[1].args[i]

# Here you've hard-coded the function name `g`. This means any other function that gets registered will be replaced by `g`
if symbol.args[0] != rhs_expr.args[0]:
    symbol = Function('g')(*dict_symbol.values())
```

You could modify `sign_defaults` in `util.py` to generate an appropriate signature. Here's the original code with additional comments for clarity

```python
def sign_defaults(symbol, expr, composition):
    '''gets defaults from an expression using composition
    
      symbol: specified by the end user. could be f(x,y,z), rho(r,theta,phi,t), etc.
      expr: the right-hand-side expression and could include any combination
        of a kamodo object's registered functions and numpy/numexpr functions
      composition: a dictionary mapping from a function symbol to its implementation
    '''
    defaults={}
    for f_ in get_undefined_funcs(expr):
        # includes {h(f(g)), f(g)}
        if str(f_) in composition:
            # ignores h(f(g))
            f_defaults = get_defaults(composition[str(f_)])
            # flatten defaults
            for arg, arg_default in f_defaults.items():
                defaults[arg] = arg_default

    arg_signatures = []
    # defaults have to go last, which may conflict with user's ordering
    # the problem is here. we need to first register the args that don't have defaults
    # after that, register the ones with defaults. hint: use continue statement
    for arg in symbol.args:
        str_arg = str(arg)
        if str_arg in defaults:
            arg_default=defaults[str(arg)]
            arg_signatures.append(forge.arg(str_arg, default=arg_default))
        else:
            arg_signatures.append(forge.arg(str_arg))

    # will raise an error if defaults are not last
    signature = forge.sign(*arg_signatures)
    return signature
```

After that, modifiy `vectorize_function` so that it calls `sign_defaults` prior to `lambdify` (read the docs on sympy's lambdify to get a handle on how it works). Instead of giving lambidfy `symbol.args`, you'll need the argument names and order returned by `sign_defaults`.

Hopefully those are the only changes. Run the tests to see if it passes. If not, you may need to go further up the call stack to `__setitem__` where the symbol was originally generated:

```python
if isinstance(lhs_expr, Symbol):
    symbol = Function(lhs_expr)(*tuple(rhs_expr.free_symbols))
```

This happens before identifying defaults in the rhs. Would need to think more about how things are ordered in `__setitem__` so these changes are easier to make going forward.

**Comment by pythonguru101** [comment]
@asher-pembroke added changes so that it registers the args that don't have defaults first and then defaults one

**Comment by asherp** [comment]
I've pulled these changes and they appear to work at first glance. However, when I look at the function signature they don't match.
```python
k = Kamodo(f=lambda x=2: x**2)
k['h'] = 'y*f' # correctly registers h(y,x)
k['g(x,y)'] = 'y*f' # should have raised an error because x appears first in signature

k.g?
```
returns
```console
Signature: k.g(y, x=2)
Docstring:
Created with lambdify. Signature:

func(x, y)
```
Todo:
1. create a test that tests both cases for h and g above, raising an error in the case for g
2. add that test to the test_kamodo.py
3. make changes to ensure that all tests pass including the new test from step 1

**Comment by pythonguru101** [comment]
added further changes, new test cases as per last feedback, all test cases passed with desired output
